http-api сервер по ТЗ 9 спринта

churmantey opened the pull request:
На ревью

churmantey committed:
http-api сервер по ТЗ 9 спринта
/tasks
/subtasks
/epics
/history
/prioritized
Добавлена библиотека gson 2.11

churmantey committed:
http-api сервер по ТЗ 9 спринта
/tasks
/subtasks
/epics
/history
/prioritized
Добавлена библиотека gson 2.11

churmantey committed:
Merge remote-tracking branch 'origin/sprint_9-solution-http-api' into sprint_9-solution-http-api

# Conflicts:
#	src/dc/yandex/kanban/service/server/handlers/SubTaskHandler.java

TarasovSam commented on src/dc/yandex/kanban/service/server/HttpTaskServer.java:
```diff
@@ -0,0 +1,73 @@
+        System.out.println("HTTP-сервер останавливается!");
+    }
+
+    public static void main(String[] args) throws IOException {
```
Понимаю, что так прописано в ТЗ
Но "смешивать кислое с соленным" - такое себе..)
Метод main все же про демонстрацию/тестирование сейчас и размещать его в классе, который относится к бизнес логике, звучит, как плохая шутка
Ну да ладно)

churmantey replied:
Для очистки совести вынес main в отдельный класс. :)

TarasovSam commented on src/dc/yandex/kanban/service/server/adapters/TaskAdapter.java:
```diff
@@ -0,0 +1,44 @@
+
+public class TaskAdapter extends TypeAdapter<Task> {
+
+    private static final DateTimeFormatter dtf = DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm");
```
У тебя ведь уже есть такая константа, ее можно импортировать)

churmantey replied:
Логично! 
Применил.

TarasovSam commented on src/dc/yandex/kanban/service/server/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,106 @@
+import java.time.Duration;
+import java.time.LocalDateTime;
+
+public abstract class BaseHttpHandler implements HttpHandler {
```
Создал общий класс для всех HTTP-обработчиков - отлично)

churmantey replied:
Подсказка была в ТЗ :)

TarasovSam commented on src/dc/yandex/kanban/service/server/adapters/LocalDateTimeAdapter.java:
```diff
@@ -0,0 +1,28 @@
+
+public class LocalDateTimeAdapter extends TypeAdapter<LocalDateTime> {
+
+    private static final DateTimeFormatter dtf = DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm");
```
Можно использовать стандартизированные типы. Например, широко используется
DateTimeFormatter.ISO_LOCAL_DATE_TIME

churmantey replied:
Спасибо, применил.

TarasovSam commented on src/dc/yandex/kanban/service/server/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,106 @@
+        h.close();
+    }
+
+    protected void sendServerError(HttpExchange h, String text) throws IOException {
```
Как вариант можно обработать ошибку IOException вместо того, чтобы везде прописывать throws IOException
Например, кинуть кастомную ошибку или напечатать ошибку отправки

churmantey replied:
Принято. Сделал вывод сообщения в консоль.

TarasovSam commented on src/dc/yandex/kanban/service/server/handlers/EpicHandler.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @Override
+    public void handlePost(HttpExchange exchange, String[] pathParts) throws IOException {
```
В POST запросах надо быть готовым, что может прийти пустое тело, это надо учесть

churmantey replied:
Отсутствие тела вроде нормально проверяется с помощью
`if (jsonElement.isJsonObject())`
В этом случае у меня выводится 404 - Передана некорректная структура эпика.
Этого недостаточно?
Или нужно какую-то отдельную ошибку/сообщение для этого варианта предусмотреть?

TarasovSam reviewed (commented):
Данила, привет! Проверил твое задание, спасибо за работу. Оставил комментарии с тем, что можно улучшить и что нужно поправить)
Жду твою работу на следующую итерацию!) Если будут вопросы, пиши в пачку)

churmantey committed:
Отдельный класс с main() для HttpTaskServer
Удалено лишнее определение DateTimeFormatter
Вместо исключений IOException при отправке ответов используется вывод сообщения в консоль.

churmantey committed:
Добавлена проверка структуры входящих json

churmantey committed:
Добавлена проверка структуры входящих json

TarasovSam commented on src/dc/yandex/kanban/service/server/adapters/TaskAdapter.java:
```diff
@@ -31,7 +28,7 @@ protected void writeCommonFields(JsonWriter jsonWriter, Task task) throws IOExce
         jsonWriter.name("status").value(task.getStatus().toString());
         if (task.getStartTime() != null) {
-            jsonWriter.name("startTime").value(task.getStartTime().format(dtf));
+            jsonWriter.name("startTime").value(task.getStartTime().format(LocalDateTimeAdapter.defaultDateTimeFormatter));
```
Да, можно через вызов класса, а можно через импорт константы)

churmantey replied:
Точно! Сделал.

TarasovSam commented on src/dc/yandex/kanban/service/server/handlers/EpicHandler.java:
```diff
@@ -73,9 +73,19 @@ public void handleGet(HttpExchange exchange, String[] pathParts) throws IOExcept
+            }
             if (jsonElement.isJsonObject()) {
                 JsonObject jsonObject = jsonElement.getAsJsonObject();
+                if (jsonObject.get("name") == null || jsonObject.get("description") == null) {
```
На будущее
Для строк нужно дополнительно проверить на пустоту и пробелы, т.е. isBlank()
name="    " пройдет текущую проверку
Сейчас таких ограничений)

churmantey replied:
Добавил проверки

TarasovSam reviewed (commented):
Данила, спасибо за то, что вносишь корректировки по рекомендациям. Дополнительных комментариев на этот раз чуть-чуть)
Задание я принимаю и до следующего спринта!) Если будут вопросы, пиши в пачку)

churmantey committed:
Общий метод проверки присутстсвия и заполненности полей переданного json
